doc/developer/design/20230418_subscribe_output.md

```diff
@@ -0,0 +1,282 @@
+bridge that gap we introduce three ways to change how subscribe
+produces its output: `WITHIN TIMESTAMP ORDER BY`, `ENVELOPE UPSERT`,
+and `ENVELOPE DEBEZIUM`. These output types are simple to implement
+because they operate on the diffs present in a single timestamp. We
+hope they'll match the client's mental model of updates on their results.
+
+# Motivation
+[motivation]: #motivation
+
+Sometimes clients want to think of the output of subscribe as a list of
+updates to the underlying relation. While the current output reflects
+the changes to the underlying relation, the lack of consistent
+ordering has troubled some clients. They worry about handling a case
+like `(k, v2, +1), (k, v1, -1)` where an update appears as an insert
+for the new value for a particular key occurs followed by the deletion
+of its previous value. The output types proposed all solve this
+problem with varying degrees of flexibility. `ENVELOPE UPSERT` will
+transform the output of `SUBSCRIBE` into a series of `UPSERTS`;
+`ENVELOPE DEBEZIUM` will transform the output into a series of
+debezium updates with a `before` and `after` state. When viewed as
+upserts the ordering of retractions and insertions doesn't
+matter. `WITHIN TIMESTAMP ORDER BY` is the most flexible: the client
+can specify a set of columns to order the output by (within each
+timestamp), including `mz_diff`. They can use this ordering to force
+retractions to occur before additions in order to simplify the client
+side handling.
+
+For the simple case of a one to one mapping between keys and values
+(there's at most one value at a time for each key), any of these
+output types would let the client see the changes to the underlying
+relation as a series of upserts. `WITHIN TIMESTAMP ORDER BY` could
+also be used to sort the results from `SUBSCRIBE`; we can think of the
+results from `SUBSCRIBE` are naturally ordered by `mz_timestamp`, we
+could supply a secondary order. I don't think there are many real use
+cases that would benefit from this.
+
+
+# Explanation
+[explanation]: #explanation
+
+`WITHIN TIMESTAMP ORDER BY` takes a `ORDER BY` expression and sorts
+the returned data within each distinct timestamp. This `ORDER BY` can
+take any column in the underlying object or query in addition to the
+`mz_diff` column. A common use case of recovering upserts to key-value
+data would be sort by all the keys and `mz_diff`. Since `SUBSCRIBE`
+guarantees that there won't be multiple updates for the same row
+(rephrased: the batch is already consolidated), this will produce data
+that looks like `(k, v1, -1), (k, v2, +1)` which are easier to handle
+for the client. If PROGRESS is set, progress messages are unaffected.
+
+`ENVELOPE UPSERT`: Takes a list of columns to interpret as a key and
+within each distinct timestamp interprets the updates as a series of
+upserts. The output columns are reordered as `(mz_timestamp, mz_state,
+k1, k2, .., v1, v2, ..)`. There is no `mz_diff` column. The values
```

Comment:
Am I understanding this correctly? There is a single row  per timestamp, and that row has (2 * k + 2) columns where k is the number of keys specified in the envelope?

Reply:
If so, what does the row look like when there are more than one value for a given key?

Reply:
added some examples to make this clearer. if a table (or select expression) has `n` columns, we'll always show those `n` columns although we reorder the columns that the user specified as key columns first